x-pack/filebeat/input/etw/input.go

```diff
@@ -0,0 +1,281 @@
+}
+
+func (op *realSessionOperator) createRealtimeSession(session *etw.Session) error {
+	return session.CreateRealtimeSession()
+}
+
+func (op *realSessionOperator) startConsumer(session *etw.Session) error {
+	return session.StartConsumer()
+}
+
+func (op *realSessionOperator) stopSession(session *etw.Session) error {
+	return session.StopSession()
+}
+
+// etwInput struct holds the configuration and state for the ETW input
+type etwInput struct {
+	log        *logp.Logger
+	config     config
+	etwSession *etw.Session
+	operator   sessionOperator
+}
+
+func Plugin() input.Plugin {
+	return input.Plugin{
+		Name:      inputName,
+		Stability: feature.Beta,
+		Info:      "Collect ETW logs.",
+		Manager:   stateless.NewInputManager(configure),
+	}
+}
+
+func configure(cfg *conf.C) (stateless.Input, error) {
+	conf := defaultConfig()
+	if err := cfg.Unpack(&conf); err != nil {
+		return nil, err
+	}
+
+	return &etwInput{
+		config:   conf,
+		operator: &realSessionOperator{},
+	}, nil
+}
+
+func (e *etwInput) Name() string { return inputName }
+
+func (e *etwInput) Test(_ input.TestContext) error {
+	return nil
+}
+
+// Run starts the ETW session and processes incoming events.
+func (e *etwInput) Run(ctx input.Context, publisher stateless.Publisher) error {
+	var err error
+
+	// Initialize a new ETW session with the provided configuration
+	e.etwSession, err = e.operator.newSession(e.config)
+	if err != nil {
+		return fmt.Errorf("error initializing ETW session: %w", err)
+	}
+
+	// Set up logger with session information
+	e.log = ctx.Logger.With("session", e.etwSession.Name)
+	e.log.Info("Starting " + inputName + " input")
+
+	// Handle realtime session creation or attachment
+	if e.etwSession.Realtime {
+		if !e.etwSession.NewSession {
+			// Attach to an existing session
+			err = e.operator.attachToExistingSession(e.etwSession)
+			if err != nil {
+				return fmt.Errorf("unable to retrieve handler: %w", err)
+			}
+			e.log.Debug("attached to existing session")
+		} else {
+			// Create a new realtime session
+			err = e.operator.createRealtimeSession(e.etwSession)
+			if err != nil {
+				return fmt.Errorf("realtime session could not be created: %w", err)
+			}
+			e.log.Debug("created session")
+		}
+	}
+	// Defer the cleanup and closing of resources
+	var wg sync.WaitGroup
+	var once sync.Once
+
+	// Create an error channel to communicate errors from the goroutine
+	errChan := make(chan error, 1)
+
+	defer func() {
+		once.Do(e.Close)
+		e.log.Info(inputName + " input stopped")
+	}()
+
+	// eventReceivedCallback processes each ETW event
+	eventReceivedCallback := func(record *etw.EventRecord) uintptr {
+		if record == nil {
+			e.log.Error("received null event record")
+			return 1
+		}
+
+		e.log.Debugf("received event %d with length %d", record.EventHeader.EventDescriptor.Id, record.UserDataLength)
+
+		data, err := etw.GetEventProperties(record)
+		if err != nil {
+			e.log.Errorf("failed to read event properties: %w", err)
```

Comment:
This verb can be used only in 'fmt.Errorf' calls (%w)

```suggestion
			e.log.Errorw("failed to read event properties", "error", err)
```